src/script/interpreter.cpp

```diff
@@ -1091,6 +1071,88 @@ class CTransactionSignatureSerializer {
     }
 };
 
+/**
+ * Wrapper that serializes like CTransaction, but with the modifications
+ *  required for the signature hash done in-place
+ */
+class CTransactionSignatureOuterSerializer {
+private:
+    const CTransaction &txTo;  //! reference to the spending transaction (the one being serialized)
+    const CScript &scriptCode; //! output script being consumed
+    const unsigned int nIn;    //! input index of txTo being signed
+    const bool fAnyoneCanPay;  //! whether the hashtype has the SIGHASH_ANYONECANPAY flag set
+    const bool fHashSingle;    //! whether the hashtype is SIGHASH_SINGLE
+    const bool fHashNone;      //! whether the hashtype is SIGHASH_NONE
+
+public:
+    CTransactionSignatureOuterSerializer(const CTransaction &txToIn, const CScript &scriptCodeIn, unsigned int nInIn, int nHashTypeIn) :
+        txTo(txToIn), scriptCode(scriptCodeIn), nIn(nInIn),
+        fAnyoneCanPay(!!(nHashTypeIn & SIGHASH_ANYONECANPAY)),
+        fHashSingle((nHashTypeIn & 0x1f) == SIGHASH_SINGLE),
+        fHashNone((nHashTypeIn & 0x1f) == SIGHASH_NONE) {}
+
+    /** Serialize the passed scriptCode, skipping OP_CODESEPARATORs */
+    template<typename S>
+    void SerializeScriptCode(S &s, int nType, int nVersion) const {
+        CScript::const_iterator it = scriptCode.begin();
+        CScript::const_iterator itBegin = it;
+        opcodetype opcode;
+        unsigned int nCodeSeparators = 0;
+        while (scriptCode.GetOp(it, opcode)) {
+            if (opcode == OP_CODESEPARATOR) {
+                nCodeSeparators++;
+            }
+        }
+        ::WriteCompactSize(s, scriptCode.size() - nCodeSeparators);
+        it = itBegin;
+        while (scriptCode.GetOp(it, opcode)) {
+            if (opcode == OP_CODESEPARATOR) {
+                s.write((char*)&itBegin[0], it-itBegin-1);
+                itBegin = it;
+            }
+        }
+        if (itBegin != scriptCode.end()) {
+            s.write((char*)&itBegin[0], it-itBegin);
+        }
+    }
+
+    /** Serialize an input of txTo */
+    template<typename S>
+    void SerializeInput(S &s, int nType, int nVersion) const {
+        // In case of SIGHASH_ANYONECANPAY, only the input being signed is serialized
+        ::WriteCompactSize(s, !fAnyoneCanPay ? 1 : 0);
+        if (!fAnyoneCanPay) {
+            // Commit to the position of the input (which
+            // is implicitly serialized in Bitcoin).
+            ::WriteCompactSize(s, nIn);
```

Comment:
[Coinspect] This line opens up a vulnerability when using fAnyoneCanPay flag. It clearly deviates from the Bitcoin CTransactionSignatureSerializer and Bitcoin segwit SIGVERSION_WITNESS_V0 hashing guarantees.
If a transaction is hashed with fAnyoneCanPay, Bitcoin hashes the txTo.vin[nIn].prevout so that OTHER inputs can be added, but the input amount and prevout consumed at index nIn cannot be modified by an attacker. This implementation allows an attacker to alter a   transaction signed with fAnyoneCanPay in a new way: replacing the signed input with a new prevout owned by the same user (using the same address) but having a different amount.  Then an attacker can wipe the user funds into fees. If the transaction also specifies fHashSingle, then the attacker can add an output and redirect all remaining bitcoins to his own address.
To fix this, replace:
::WriteCompactSize(s, nIn);
With:
:Serialize(s, txTo.vin[nIn].prevout, nType, nVersion);
This will allow an attacker to change the index of the input in the transaction input list (as in Bitcoin), but does not open the new vector of attack.
Note that hashing the input amount is not enough, as the attacker can choose another unspent output with the same amount belonging to the same user and confuse the wallet that generated the transaction.